docs/internal/abi.md

```diff
@@ -0,0 +1,437 @@
+The tables will be compacted (either during colouring or afterwards) to minimise that.
+
+## Union Types
+
+Union types need as many representations as there are types in the union.
+Objects of union types need to be able to represent all types as well as identify which one for each instance.
+
+Example:
+```ts
+// Must handle both U8 and F32
+// Must also know which one at run time
+function(arg : U8 | F32) { ... }
+...
+var a : U8 & imm = 10;
+function(a); // Passes a 1-byte integer
+var b : F32 & imm = 3.14;
+function(b); // Passes 32-bit floating point
+```
+
+All objects are referred to by pointers.
+Machine-word types are refereed to by value as an optimisation by the compiler.
+
+For that reason, they can only _contain_ the following representations:
+* Machine-word values.
+* Pointers to objects.
+
+To identify which type the run-time object has, we need a discriminator flag.
+Because the values take up all their storage, we need a wider storage to bundle more than one type in the same representation.
+
+Objects have a descriptor that identifies their types, so pointers don't need a special descriptor in the union representation.
+So a union of pointers is represented as just a pointer and the discrimination will happen at the object level.
+However, numeric types don't have discriminators, so they need special handling when inside union.
+Furthermore, mixing pointers and numeric types creates the need to differentiate pointers from the rest.
+
+The type representation will be composed of a _descriptor_ and a _payload_ with the size of the largest object.
+There are two ways of storing the descriptor: `wide packing` and `NaN-boxing`.
+
+### Wide packing
+
+This is the naive representation, using an 8-bit descriptor and the largest object's size as the payload.
+
+Example:
+```ts
+(A | U64) -> (Pointer | i64) -> { i8, i64 } // On both 32-bit and 64-bit machines
+// On CHERI, this would be { i8, i128 }, as pointers are 128-bits wide
+```
+
+This can be used for all types, but it creates two main problems:
+1. It adds at least 8-bits to every number.
+   For a `U8`, that's twice the size.
+2. Alignment requirements may force us to _pad_ the first `i8`.
+   This would add another 24, 56 or 120 bits on structures or arrays, per element.
+
+For this reason, we use NaN-boxing below for all types that we can.
+The encoding of the discriminator is shown below.
+
+### NaN-Boxing
+
+IEEE-754 doubles are 64 bits, on both 32-bit and 64-bit platforms, encoded in the exact same way:
+```
+[ sign ][ exponent ][ mantissa ]
+63     62          51          0
+```
+So:
+* 1 sign bit
+* 11 exponent bits
+* 52 mantissa bits
+
+A `NaN` (not-a-number) is encoded in IEEE-754 as setting the exponent to _all ones_.
+All other bits are ignored.
+If we set the exponent to _all ones_, we can use the remaining 53 bits to encode a variety of smaller types.
```

Comment:
There's a little bit of subtlety here around qNaNs vs sNaNs, but it's probably fine for now.

Reply:
Right, so I'm assuming we don't care about that because we're ignoring it in the boxing and we don't want any traps or signals in the run time, so all NaNs are quiet?